### PR TITLE
[Concurrency] TaskLocal.withValue closure should produce no warnings inside Actor

### LINKED ISSUE #62220: Actor boundary warning for @TaskLocal values
**Description**
I am getting this warning with Version 14.1 (14B47b):

```swift
actor Test {

	@TaskLocal static var local: Int?

	func run() async {
		await Self.$local.withValue(42) {  // ⚠️
			await work()
		}
	}

	func work() async {
		print("Hello \(Self.local ?? 0)")
	}
}

await Test().run()
```

> warning: non-sendable type '() async throws -> ()' exiting actor-isolated context in call to non-isolated instance method 'withValue(_:operation:file:line:)' cannot cross actor boundary

**Steps to reproduce**
1. copy the above code into a file `test.swift`
2. run `swift -warn-concurrency test.swift`

**Expected behavior**
The code should not produce a warning if it is legal code.

**Environment**
* swift-driver version: 1.62.15 Apple Swift version 5.7.1 (swiftlang-5.7.1.135.3 clang-1400.0.29.51)
* Target: x86_64-apple-macosx13.0
* Xcode 14.1, Build version 14B47b

**Additional context**
Mentioned on [Swift forum](https://forums.swift.org/t/actor-boundary-warning-for-tasklocal-values/59499).

### PR DESCRIPTION
**Pull request by ktoso**
Without the @Sendable, using a task-local's `withValue` results with unexpected warnings which scared developers into thinking they might be misusing the API.

This PR adds the missing annotation; I believe this is ABI compatible, but would like to make sure.

Resolves https://github.com/apple/swift/issues/62220
rdar://102638772

### CODE REVIEW

**Comment by ktoso** [comment]
resolves rdar://102638772

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/BackDeployConcurrency/TaskLocal.swift [hunk 1]
```diff
@@ -134,8 +134,9 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
   ///
   /// If the value is a reference type, it will be retained for the duration of
   /// the operation closure.
+  @Sendable
```
Comment (pyrtsa) [inline stdlib/public/BackDeployConcurrency/TaskLocal.swift:137]: Why is the back-deployed `withValue` itself also marked as `@Sendable` but not the `withValue` implementation in stdlib/public/Concurrency/TaskLocal.swift?

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by rjmccall** [comment]
I think the right fix here is that this ought to inherit the caller's executor.

**Comment by kavon** [comment]
> I believe this is ABI compatible, but would like to make sure.

Seems it is _not_ ABI compatible to add `@Sendable` to a parameter. For example:

```
> swift demangle s4what8whatever2fnS3iXE_tYaF
$s4what8whatever2fnS3iXE_tYaF ---> what.whatever(fn: (Swift.Int) -> Swift.Int) async -> Swift.Int
> swift demangle s4what8whatever2fnS3iYbXE_tYaF
$s4what8whatever2fnS3iYbXE_tYaF ---> what.whatever(fn: @Sendable (Swift.Int) -> Swift.Int) async -> Swift.Int
```

**Comment by ktoso** [comment]
Ah the unsafe executor inheritance... I'll give that a shot.

Thanks for chiming in @rjmccall, @kavon!

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by rjmccall** [comment]
I think it's correct in the abstract to make this inherit the executor.  However, marking the function as inheriting the executor makes a semantic guarantee that the function won't switch, and unfortunately the existing implementations of this function (ever since 5.6 / SE-0338) *do* switch (to the generic executor) (unless that switch has been optimized out in the release builds).  I think it's okay for existing compilations to call a function that doesn't switch (the function doesn't *rely* on being on the generic executor, and you can think of turning off the switches as manually optimizing them away, which is permitted).  However, it's not okay for new compilations that will rely on the absence of switching to potentially call a function that does switch.  So we need to have a way to back-deploy this fix for it to be correct.

It looks like everything these functions call is back-deployable — they're either already `@usableFromInline` or they're runtime functions already exported to which we can safely add the attribute.  So we need new compilations to call the new version of the function, but also emit a stable version of the function for existing users.  @DougGregor points out that `@backDeployed` is exactly what we want for this — it's a lie only in that the function does already exist, but since we don't want to actually use the existing function, it's a pretty consequence-free lie.  We can make the function `@inlinable` as well for performance.

**Comment by ktoso** [comment]
Thank you for the analysis and guidance, John!

We'll need to wait for `_backDeploy` to be able to be combined with inlinable, but that's coming soon. I'll update the PR and wait for it :)

**Comment by ktoso** [comment]
Depends on https://github.com/apple/swift/pull/62309

**Comment by tshortli** [comment]
Adopting `@_backDeploy` here should be unblocked by a combination of https://github.com/apple/swift/pull/62444 and https://github.com/apple/swift/pull/62427.

**Comment by ktoso** [comment]
Great, thank you!

**Comment by ktoso** [comment]
Ok those got merged, so I'll try to revive this 👍

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 2]
```diff
@@ -102,6 +102,7 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
     self.defaultValue = defaultValue
   }
 
+  @usableFromInline
```
Comment (DougGregor) [inline stdlib/public/Concurrency/TaskLocal.swift:105]: `@usableFromInline` adds a public symbol that won't be available on older platforms. Can you make this `@alwaysEmitIntoClient`?

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 3]
```diff
@@ -135,6 +136,9 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
   /// If the value is a reference type, it will be retained for the duration of
   /// the operation closure.
   @discardableResult
+  @_unsafeInheritExecutor
+  @available(SwiftStdlib 5.1, *) // back deploy requires we declare the availability explicitly on this method
+  @_backDeploy(before: macOS 9999, iOS 9999, watchOS 9999, tvOS 9999) // SwiftStdlib 5.8 but it doesn't work with _backDeploy
```
Comment (DougGregor) [inline stdlib/public/Concurrency/TaskLocal.swift:141]: How sad. @tshortli can `@backDeploy` support `SwiftStdlib 5.8` here?

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 4]
```diff
@@ -160,6 +164,7 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
   /// If the value is a reference type, it will be retained for the duration of
   /// the operation closure.
   @discardableResult
+  @inline(__always)
```
Comment (DougGregor) [inline stdlib/public/Concurrency/TaskLocal.swift:167]: Why do we need this?

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 5]
```diff
@@ -212,23 +218,27 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
 // ==== ------------------------------------------------------------------------
 
 @available(SwiftStdlib 5.1, *)
+@usableFromInline
```
Comment (DougGregor) [inline stdlib/public/Concurrency/TaskLocal.swift:221]: Hmm. Same ABI issue here, this is adding a public symbol that wasn't there before.

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 3 above)
Comment (tshortli) [reply stdlib/public/Concurrency/TaskLocal.swift:141]: `@_backDeploy` supports availability macros in general (they're heavily used in some of the tests). If there's an edge case exposed here, though, let me know @ktoso

Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift:141]: Yeah it didn’t work with the SwiftStdlib and I tried looking into it but didn’t quite see how those get applied. Help very welcome on this @tshortli !

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 4 above)
Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift:167]: Was John’s suggestion to add those

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 2 above)
Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift:105]: Ok will do, I hope backdeploy will understand that hmm.

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 4 above)
Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift:167]: I over did perhaps, removing all except the one we need.

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 5 above)
Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift:221]: I'm confused about this one, isn't this effectively "call the swift_task_localValuePush" and the symbol already was there?

I'll change to `@alwaysEmitIntoClient` as you suggested but don't understand why, could you explain a bit?

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 2 above)
Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift:105]: Okey cool it does :)

File: stdlib/public/BackDeployConcurrency/TaskLocal.swift [hunk 6]
```diff
@@ -212,13 +218,15 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
 // ==== ------------------------------------------------------------------------
 
 @available(SwiftStdlib 5.1, *)
+@_alwaysEmitIntoClient
 @_silgen_name("swift_task_localValuePush")
```
Comment (ktoso) [inline stdlib/public/BackDeployConcurrency/TaskLocal.swift:222]: hope this is right now then @DougGregor ?

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 7]
```diff
@@ -134,7 +135,11 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
   ///
   /// If the value is a reference type, it will be retained for the duration of
   /// the operation closure.
+  @inlineable
   @discardableResult
+  @_unsafeInheritExecutor
+  @available(SwiftStdlib 5.1, *) // back deploy requires we declare the availability explicitly on this method
+  @_backDeploy(before: SwiftStdlib 5.8)
```
Comment (ktoso) [inline stdlib/public/Concurrency/TaskLocal.swift:142]: Yay, it works actually.

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 4 above)
Comment (ktoso) [reply stdlib/public/Concurrency/TaskLocal.swift]: Just added inlinable to the `with...` methods now, I think that's what @rjmccall had in mind?

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 8]
```diff
@@ -212,13 +218,15 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
 // ==== ------------------------------------------------------------------------
 
 @available(SwiftStdlib 5.1, *)
+@_alwaysEmitIntoClient
```
Comment (DougGregor) [inline stdlib/public/Concurrency/TaskLocal.swift:221]:
```suggestion
@usableFromInline
```

File: stdlib/public/Concurrency/TaskLocal.swift [hunk 9]
```diff
@@ -212,13 +218,15 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
...
 @_silgen_name("swift_task_localValuePush")
 func _taskLocalValuePush<Value>(
   key: Builtin.RawPointer/*: Key*/,
   value: __owned Value
 ) // where Key: TaskLocal
 
 @available(SwiftStdlib 5.1, *)
+@_alwaysEmitIntoClient
```
Comment (DougGregor) [inline stdlib/public/Concurrency/TaskLocal.swift:229]:
```suggestion
@usableFromInline
```

File: stdlib/public/BackDeployConcurrency/TaskLocal.swift (hunk 6 above)
Comment (DougGregor) [reply stdlib/public/BackDeployConcurrency/TaskLocal.swift:222]: Should be `@usableFromInline`.

File: stdlib/public/BackDeployConcurrency/TaskLocal.swift [hunk 10]
```diff
@@ -212,13 +218,15 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
 // ==== ------------------------------------------------------------------------
 
 @available(SwiftStdlib 5.1, *)
+@_alwaysEmitIntoClient
```
Comment (DougGregor) [inline stdlib/public/BackDeployConcurrency/TaskLocal.swift:221]:
```suggestion
@usableFromInline
```

File: stdlib/public/Concurrency/TaskLocal.swift (hunk 7 above)
Comment (DougGregor) [reply stdlib/public/Concurrency/TaskLocal.swift:142]: w00t!

**Review by DougGregor — APPROVED** [review]
Looks great now!

File: stdlib/public/BackDeployConcurrency/TaskLocal.swift [hunk 11]
```diff
@@ -212,13 +218,15 @@ public final class TaskLocal<Value: Sendable>: Sendable, CustomStringConvertible
...
 @_silgen_name("swift_task_localValuePush")
 func _taskLocalValuePush<Value>(
   key: Builtin.RawPointer/*: Key*/,
   value: __owned Value
 ) // where Key: TaskLocal
 
 @available(SwiftStdlib 5.1, *)
+@_alwaysEmitIntoClient
```
Comment (ktoso) [inline stdlib/public/BackDeployConcurrency/TaskLocal.swift:229]:
```suggestion
@usableFromInline
```

**Comment by ktoso** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
@swift-ci please smoke test and merge

**Pull request merged**